sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuPartitioning.scala

```diff
@@ -195,47 +195,60 @@ trait GpuPartitioning extends Partitioning {
     }
   }
 
-  def sliceAndSerializeOnGpu(numRows: Int, partitionIndexes: Array[Int],
+  private def gpuSplitAndSerialize(table: Table, slices: Int*): Array[DeviceMemoryBuffer] = {
+    NvtxRegistry.GPU_KUDO_SERIALIZE {
+      KudoGpuSerializer.splitAndSerializeToDevice(table, slices: _*)
+    }
+  }
+
+  private def sliceAndSerializeOnGpu(numRows: Int, partitionIndexes: Array[Int],
       partitionColumns: Array[GpuColumnVector]): Array[(ColumnarBatch, Int)] = {
     partitionColumns.foreach(_.getBase.getNullCount)
     val (dataHost, offsetsHost) = withResource(partitionColumns) { _ =>
       withResource(new Table(partitionColumns.map(_.getBase).toArray: _*)) { table =>
-        withResource(KudoGpuSerializer.splitAndSerializeToDevice(table,
+        withResource(gpuSplitAndSerialize(table,
           partitionIndexes.tail: _*)) { dmbs =>
           val data = dmbs(0)
           val offsets = dmbs(1)
           closeOnExcept(Seq(HostMemoryBuffer.allocate(data.getLength),
             HostMemoryBuffer.allocate(offsets.getLength))) { seq =>
             val dataHost = seq(0)
             val offsetsHost = seq(1)
-            dataHost.copyFromDeviceBuffer(data, Cuda.DEFAULT_STREAM)
-            offsetsHost.copyFromDeviceBuffer(offsets, Cuda.DEFAULT_STREAM)
+            NvtxRegistry.GPU_KUDO_COPY_TO_HOST {
+              dataHost.copyFromDeviceBuffer(data, Cuda.DEFAULT_STREAM)
+              offsetsHost.copyFromDeviceBuffer(offsets, Cuda.DEFAULT_STREAM)
+              Cuda.DEFAULT_STREAM.sync()
```

Comment:
Was the lack of synchronization we had before causing an illegal access?

In this case, we may want to issue all the copies, then synchronize, then free.

Reply:
There was no illegal access, I think I just forgot to make the copies async